platforms/core-runtime/collections/src/main/java/org/gradle/internal/collect/ArrayCopy.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2025 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.collect;
+
+import java.util.Arrays;
+
+/// Different array copying operations required by the persistent collection
+/// implementations.
+///
+/// In general, these should be as efficient as possible, contain
+/// no branching code, and minimize [System] calls.
+/// In practice, it gets [complicated][#insertAtPushingRight(int, Object\[\], Object\[\], int, int, int)].
+final class ArrayCopy {
+
+    static final Object[] EMPTY_ARRAY = new Object[0];
+
+    static Object[] append(Object[] array, Object newElement) {
+        Object[] newArray = Arrays.copyOf(array, array.length + 1);
+        newArray[array.length] = newElement;
+        return newArray;
+    }
+
+    static Object[] append(Object[] array, Object e1, Object e2) {
+        int length = array.length;
+        Object[] newArray = Arrays.copyOf(array, length + 2);
+        newArray[length] = e1;
+        newArray[length + 1] = e2;
+        return newArray;
+    }
+
+    static Object[] replaceAt(int index, Object[] array, Object newElement) {
+        Object[] newArray = Arrays.copyOf(array, array.length);
+        newArray[index] = newElement;
+        return newArray;
+    }
+
+    static Object[] insertAt(int index, Object[] array, Object newElement) {
+        Object[] newArray = Arrays.copyOf(array, array.length + 1);
+        System.arraycopy(array, index, newArray, index + 1, array.length - index);
+        newArray[index] = newElement;
+        return newArray;
+    }
+
+    static Object[] insertAt(int index, Object[] array, Object e1, Object e2) {
+        Object[] newArray = Arrays.copyOf(array, array.length + 2);
+        System.arraycopy(array, index, newArray, index + 2, array.length - index);
+        newArray[index] = e1;
+        newArray[index + 1] = e2;
+        return newArray;
+    }
+
+    // 🤔 Note: This has different behavior based on payload (0 vs 1).
```

Comment:
Seems like this should be two different methods then? Why is payload an integer if we only allow 0 and 1?